File: WdtcCore/src/main/java/org/wdt/wdtc/download/quilt/QuiltInstallTask.java
package org.wdt.wdtc.download.quilt;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import org.apache.log4j.Logger;
import org.wdt.platform.DependencyDownload;
import org.wdt.platform.gson.JSONArray;
import org.wdt.platform.gson.JSONObject;
import org.wdt.platform.gson.JSONUtils;
import org.wdt.wdtc.download.DownloadTask;
import org.wdt.wdtc.download.infterface.DownloadSource;
import org.wdt.wdtc.download.infterface.InstallTask;
import org.wdt.wdtc.game.GameVersionJsonObject;
import org.wdt.wdtc.game.Launcher;
import org.wdt.wdtc.game.LibraryObject;
import org.wdt.wdtc.utils.WdtcLogger;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class QuiltInstallTask extends QuiltDownloadInfo implements InstallTask {

    private static final Logger logmaker = WdtcLogger.getLogger(QuiltInstallTask.class);
    private final DownloadSource source;

    public QuiltInstallTask(Launcher launcher, String quiltVersionNumber) {
        super(launcher, quiltVersionNumber);
        this.source = Launcher.getDownloadSource();
    }


    public void DownloadQuiltGameVersionJson() {
        DownloadTask.StartDownloadTask(getQuiltVersionJsonUrl(), getQuiltVersionJson());
    }


    @Override
    public void execute() throws IOException {
        GameVersionJsonObject VersionJsonObject = launcher.getGameVersionJsonObject();
        JSONObject QuiltVersionJsonObject = getQuiltGameVersionJsonObject();
        JSONObject Arguments = QuiltVersionJsonObject.getJSONObject("arguments");
        GameVersionJsonObject.Arguments arguments = VersionJsonObject.getArguments();
        JsonArray GameList = arguments.getGameList();
        VersionJsonObject.setMainClass(QuiltVersionJsonObject.getString("mainClass"));
        GameList.addAll(Arguments.getJSONArray("game").getJsonArrays());
        arguments.setGameList(GameList);
        VersionJsonObject.setArguments(arguments);
        JSONArray QuiltLibraryList = QuiltVersionJsonObject.getJSONArray("library");
        List<LibraryObject> LibraryList = VersionJsonObject.getLibraries();
        for (int i = 0; i < QuiltLibraryList.size(); i++) {
            JSONObject QuiltLibraryObject = QuiltLibraryList.getJSONObject(i);
            DependencyDownload download = new DependencyDownload(QuiltLibraryObject.getString("name"));
            String LibraryDefaultUrl = QuiltLibraryObject.getString("url");
            if (LibraryDefaultUrl.equals("https://maven.fabricmc.net/")) {
                download.setDefaultUrl(source.getFabricLibraryUrl());
            } else if (LibraryDefaultUrl.equals("https://maven.quiltmc.org/repository/release/")) {
                download.setDefaultUrl("https://maven.quiltmc.org/repository/release/");
            }
            LibraryList.add(LibraryObject.getLibraryObject(download, LibraryDefaultUrl));
        }
        VersionJsonObject.setLibraries(LibraryList);
        VersionJsonObject.setId(launcher.getVersionNumber() + "-quilt-" + QuiltVersionNumber);
        launcher.PutToVersionJson(VersionJsonObject);
    }

    @Override
    public void setPatches() throws IOException {
        GameVersionJsonObject Object = launcher.getGameVersionJsonObject();
        List<JsonObject> ObjectList = new ArrayList<>();
        ObjectList.add(JSONUtils.getJsonObject(launcher.getVersionJson()));
        ObjectList.add(JSONUtils.getJsonObject(getQuiltVersionJson()));
        Object.setJsonObject(ObjectList);
        launcher.PutToVersionJson(Object);
    }

    @Override
    public void AfterDownloadTask() {

    }

    @Override
    public void BeforInstallTask() {
        DownloadQuiltGameVersionJson();
    }
}

File: core/core-impl/src/main/kotlin/org/wdt/wdtc/core/impl/download/quilt/QuiltInstallTask.kt
package org.wdt.wdtc.core.impl.download.quilt

import org.wdt.utils.gson.*
import org.wdt.wdtc.core.openapi.download.interfaces.ModInstallTaskInterface
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsJsonObjectInterface
import org.wdt.wdtc.core.openapi.game.Version
import org.wdt.wdtc.core.openapi.utils.STRING_SPACE
import org.wdt.wdtc.core.openapi.utils.cleanStrInString
import org.wdt.wdtc.core.openapi.utils.startDownloadTask
import java.io.IOException

class QuiltInstallTask : QuiltDownloadInfo, ModInstallTaskInterface {
	
	constructor(version: Version, quiltVersionNumber: String) : super(version, quiltVersionNumber)
	
	constructor(version: Version, versionsJsonObjectInterface: VersionsJsonObjectInterface) : super(
		version,
		versionsJsonObjectInterface
	)
	
	override suspend fun beforInstallTask() {
		startDownloadTask(quiltVersionJsonUrl, quiltVersionJson)
	}
	
	override suspend fun overwriteVersionJson() {
		quiltProfileJsonObject.also { quilt ->
			version.run {
				gameVersionJsonObject.apply {
					gameVersionJsonObject.apply {
						id = "${versionNumber}-quilt-$modVersion"
						mainClass = quilt.mainClass
						arguments = arguments.apply {
							jvmList = jvmList.also { game ->
								quilt.arguments.jvmList?.onEach {
									game!!.add(it.asString.cleanStrInString(STRING_SPACE))
								}
							}
							gameList = gameList.also { game ->
								quilt.arguments.gameList?.onEach {
									game!!.add(it.asString)
								}
							}
						}
						libraries.addAll(quilt.libraries.map {
							it.toLibraryObject()
						})
					}.also {
						it.putToVersionJson()
					}
					
				}
			}
		}
	}
	
	@Throws(IOException::class)
	override fun writeVersionJsonPatches() {
		version.run {
			gameVersionJsonObject.apply {
				patches = mutableListOf(
					versionJson.readFileToJsonObject(),
					quiltVersionJson.readFileToJsonObject()
				)
			}.putToVersionJson()
		}
	}
	
	override suspend fun afterDownloadTask() {}
}
